### PR TITLE
sample: ipc: Fix openamp rsc table sample & documentation

### PR DESCRIPTION
**Pull request by Marquichou**
Hello everyone,

While testing the openamp rsc table sample I noticed a little glitch in the Linux logs for the TTY responder as shown down below.

```
TTY 0x0401: Hello Zephyr�
```

The zephyr sample was sending a buffer with one extra character compared to what is supposed to be sent.

```
TTY 0x0401: Hello Zephyr
```

Then I compare the logs from both Linux & Zephyr while executing the sample to what the documentation shows and I notice that the documentation was not accurate. Thus I decided to add logs to stick with it and update the documentation to reflect reality.

It was tested on a Renesas spider board while developing it's (downstream) openamp driver !

### CODE REVIEW

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 1]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
 the co-processor.
 
 Building the application
-*************************
+************************
 
 Zephyr
--------
+------
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:24]: This should actually be `=` not `-` as per https://docs.zephyrproject.org/latest/contribute/documentation/guidelines.html#headings -- would be great if you could perform the mass replace :)

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 2]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
+*************************
 
-Zephyr console
----------------
+Linux setup
+-----------
 
-Open a serial terminal (minicom, putty, etc.) and connect the board with the
-following settings:
+Enable SAMPLE_RPMSG_CLIENT configuration to build the rpmsg_client_sample.ko module.
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:36]:
```suggestion
Enable ``SAMPLE_RPMSG_CLIENT`` configuration to build the :file:`rpmsg_client_sample.ko` module.
```

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 3]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
 
-Reset the board.
+Open a serial terminal (minicom, putty, etc.) and connect to the board using default serial port settings.
 
 Linux console
----------------
+-------------
 
-Open a Linux shell (minicom, ssh, etc.) and insert a module into the Linux Kernel
+Open a Linux shell (minicom, ssh, etc.) and insert the rpmsg_client_sample module into the Linux Kernel.
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:46]:
```suggestion
Open a Linux shell (minicom, ssh, etc.) and insert the ``rpmsg_client_sample`` module into the Linux Kernel.
```

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 4]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
+Zephyr setup
+------------
+
+Open a serial terminal (minicom, putty, etc.) and connect to the board using default serial port settings.
+
+Linux console
+-------------
+
+Open a Linux shell (minicom, ssh, etc.) and print the messages coming through the rpmsg-tty endpoint created during the sample initialization.
+On the Linux console, send a message to Zephyr which answers with the "TTY <add>" prefix.
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:103]:
```suggestion
On the Linux console, send a message to Zephyr which answers with the :samp:`TTY {<add>}:` prefix.
```

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 5]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
 .. code-block:: console
 
    $> cat /dev/ttyRPMSG0 &
    $> echo "Hello Zephyr" >/dev/ttyRPMSG0
    TTY 0x0401: Hello Zephyr
+
+Zephyr console
+--------------
+
+On the zephyr console the received message is displayed as shown down below then sent back to Linux.
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:115]:
```suggestion
On the Zephyr console. the received message is displayed as shown below, then sent back to Linux.
```

**Review by kartben — CHANGES_REQUESTED** [review]
Thanks for the PR, much appreciated!

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 1 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Thanks for the review, done !

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 2 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Thanks for the review, done !

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 3 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Thanks for the review, done !

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 4 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Thanks for the review, done !

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 5 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Thanks for the review, done !

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 6]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
+   OpenAMP Linux sample client responder ended
+
+
+Running the rpmsg TTY demo
+**************************
+
+Linux setup
+===========
 
+Enable RPMSG_TTY in the kernel configuration.
```
Comment (kartben) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:92]: forgot this little guy

```suggestion
Enable ``RPMSG_TTY`` in the kernel configuration.
```

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 4 above)
Comment (kartben) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: the curly braces were intentional :) (see https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-samp) also looks like it should actually be "addr" not "add"  to be consistent with the line below.

**Review by kartben — CHANGES_REQUESTED** [review]
Thanks for quickly implementing the fixes! See a couple additional comments. Also, in case you didn't know, you may check the CI-built page here: https://builds.zephyrproject.io/zephyr/pr/66426/docs/samples/subsys/ipc/openamp_rsc_table/README.html

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 4 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: Oh sorry for the missing curly braces ! I also switch from "add" to "addr" as you mentioned.

File: samples/subsys/ipc/openamp_rsc_table/README.rst (hunk 6 above)
Comment (Marquichou) [reply samples/subsys/ipc/openamp_rsc_table/README.rst]: done !

**Comment by Marquichou** [comment]
I notice that, when executing an rpmsg TTY communication after the client sample, I had incorrect messages displayed in the Zephyr logs I added in this pull request.

In Linux the message that is sent is also the received message, so everything looks good.

```bash
echo test > /dev/ttyRPMSG0
[root@localhost ~]# TTY 0x0401: test
```

However in Zephyr, the following message is displayed.

```bash
[Linux TTY] incoming msg: test
 world!
```

So it seems that mailboxes are not cleaned between communication and so the way I wrote the printk, without specifying the length of the message, is incorrect. I correct it by doing:

```patch
diff --git a/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c b/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
index 427c499e1a..2b25fb3d7a 100644
--- a/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
+++ b/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
@@ -291,7 +291,7 @@ void app_rpmsg_client_sample(void *arg1, void *arg2, void *arg3)
        while (msg_cnt < 100) {
                k_sem_take(&data_sc_sem,  K_FOREVER);
                msg_cnt++;
-               printk("[Linux sample client] incoming msg %d: %s\n", msg_cnt, (char *)sc_msg.data);
+               printk("[Linux sample client] incoming msg %d: %.*s\n", msg_cnt, sc_msg.len, (char *)sc_msg.data);
                rpmsg_send(&sc_ept, sc_msg.data, sc_msg.len);
        }
        rpmsg_destroy_ept(&sc_ept);
@@ -320,7 +320,7 @@ void app_rpmsg_tty(void *arg1, void *arg2, void *arg3)
        while (tty_ept.addr !=  RPMSG_ADDR_ANY) {
                k_sem_take(&data_tty_sem,  K_FOREVER);
                if (tty_msg.len) {
-                       printk("[Linux TTY] incoming msg: %s", (char *)tty_msg.data);
+                       printk("[Linux TTY] incoming msg: %.*s", tty_msg.len, (char *)tty_msg.data);
                        snprintf(tx_buff, 13, "TTY 0x%04x: ", tty_ept.addr);
                        memcpy(&tx_buff[12], tty_msg.data, tty_msg.len);
                        rpmsg_send(&tty_ept, tx_buff, tty_msg.len + 12);
```

**Comment by aaillet** [comment]
Hey @kartben ,
Is current status of the sample documentation good for you ?

File: samples/subsys/ipc/openamp_rsc_table/README.rst [hunk 7]
```diff
@@ -18,73 +18,110 @@ a Linux kernel OS on the main processor and a Zephyr application on
...
+============
 
-On the Linux console send a message to Zephyr which answers with the "TTY <add>" prefix.
+Open a serial terminal (minicom, putty, etc.) and connect to the board using default serial port settings.
+
+Linux console
+=============
+
+Open a Linux shell (minicom, ssh, etc.) and print the messages coming through the rpmsg-tty endpoint created during the sample initialization.
+On the Linux console, send a message to Zephyr which answers with the :samp:`TTY {<addr>}:` prefix.
```
Comment (arnopo) [inline samples/subsys/ipc/openamp_rsc_table/README.rst:103]: Regarding the text generated by GitHub, seems that `:samp:` is not well interpreted (https://github.com/iotbzh/zephyr/tree/fix_openamp_sample/samples/subsys/ipc/openamp_rsc_table#id6)

Is there a way to make it working for both GitHub and sphinx?

Comment (kartben) [reply samples/subsys/ipc/openamp_rsc_table/README.rst:103]: I doubt it. GitHub only has basic ReStructuredText support so it's quite expected that the GitHub preview is not fully interpreted. See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/getting_started/index.rst

**Review by kartben — APPROVED** [review]

**Comment by kartben** [comment]
> Hey @kartben , Is current status of the sample documentation good for you ?

Yes, thanks! LGTM but someone actually knowledgeable about the topic like @carlocaione and @arnopo will need to confirm the changes are also good beyond the cosmetic/style aspect :) 
Thanks for your work on this, very much appreciated!

**Review by arnopo — APPROVED** [review]
LGTM, thanks for the updates!

**Review by carlocaione — APPROVED** [review]

**Pull request merged**